### ISSUE TITLE
Add Neighbourhood Algorithm (maybe in 3rd-party package?)

### ISSUE THREAD

**Issue by benbovy** [scikit-optimize/scikit-optimize#676]
I'm interested to implement the Neighbourhood Algorithm (NA) in Python. This algorithm is quite popular in the geophysical community for solving optimization/inversion problems. It is detailed in [this paper](https://academic.oup.com/gji/article/138/2/479/596234#32272598) and I summarize it here below.

<details>
<summary><b>Click for description of the Neighbourhood Algorithm</b></summary>
<p>
NA is an algorithm that can be used for gradient-free global optimization. It is based on Voronoï tessellation of the n-d parameter space. After having evaluated the objective function at random points in the parameter space at the first iteration, each next iteration consists of:

- building a simple model of the objective function in the parameter space, which consists of a Voronoï diagram drawn from all the points that have been generated so far (building the whole diagram is actually not needed, but saying so is conceptually valid).
- generating `ns` new points within the `nr` "best fit" Voronoi cells (i.e. `ns `/`nr` are generated uniformly in each of the those cells)
- evaluate the objective function for those newly generated points

The algorithm has thus two tuning parameters: `ns` and `nr`.
</p>
</details>
<br>

Instead of reinventing the wheel, it would be great to reuse the nice features that are implemented in scikit-optimize like its API, high level plotting functions, benchmarks, (dask)/joblib integration, serialization of optimization results and possibly more in a near future.

However, I'm not sure if the algorithm described above would fit in scikit-optimize. 

<details>
<summary><b>Potential issues</b></summary>
<p>

- Usually we run this algorithm with a lot of forward model runs (e.g.,  in the order of 1e5), whereas (from what I understand) the current algorithms in this library rather focus on few evaluations of costly objective functions. Would `skopt` easily support this amount evaluations? 
- We mainly use NA because there is potential to run (many) forward models in parallel at each iteration. I guess this could be a-priori implemented with the ask and tell interface of `Optimizer`, although in this case the strategy for asking new points seems different than any of the available strategies (not sure if the constant liar approach is compatible).
- I'm not very familiar with `sklearn` and how it is used here so I don't know if it is feasible to implement this algorithm within this framework. 

</p>
</details>
<br>

Also, I would fully understand if you feel that such algorithm has not its place in`skopt`. My question is then: is `skopt` designed so that it is easy to extend it with additional algorithms while still leveraging all the features mentioned above? For example, is it safe to subclass `Optimizer` or create my own class with a similar interface?

Sorry for this long post and for asking a lot of questions. I'm very new to `skopt`. I should probably have looked more into `skopt`'s docs and code before, but I really want to avoid unnecessary development effort if I can just reuse (at least) bits from `skopt`. If you think this algorithm would fit in scikit-optimize, then I'm happy to start working on a PR.

**Comment by benbovy** [comment]
I should add that this algorithm has been published a while ago, and although I'm not very familiar with the most advanced methods currently implemented in skopt I wouldn't be surprised if the later outperform it in many aspects.

IMO our community (geophysics / geodynamics / geomorphology) would still benefit from an implementation of NA within the Python scientific ecosystem (a Fortran version already exists). This would be useful at least for quick comparison/benchmarking and for legacy purpose.

**Comment by glouppe** [comment]
Hi Benoit! Good to see you around here.

There isn't much activity these days, mostly because the few core devs are busy their life, so it might take some time for us to respond.

In principles, I think it would be nice that scikit-optimize eventually expands beyond bayesian optimization to include more gradient-free optimization techniques. However, every feature we include adds in terms of maintenance cost, which we can already barely sustain.

On the other hand, developing a new algorithm might help us identify issues in the interface with regards to use cases we didn't really consider, so that might be beneficial in the end.

What do others think? @betatim @iaroslav-ai 

@benbovy Besides the interface, do you see components that could be readily reused?

**Comment by benbovy** [comment]
Hi Gilles! Thanks for your answer. No worries if it takes time for you to respond, that perfectly fine!

> every feature we include adds in terms of maintenance cost, which we can already barely sustain.

Yeah, I fully understand that and I'm fine with the idea of implementing it in a separate package.

> Besides the interface, do you see components that could be readily reused?

Most of the components I think:

- ``skopt.space``
- ``skopt.plots``
- ``skopt.callbacks``
- ``skopt.benchmarks``

From a less technical point of view, my hope with integrating NA with scikit-optimize is also "advertising" the optimization methods already implemented in ``skopt``, which would be more discoverable to our community (including me) and perhaps better suited depending on the problem we want to solve.

> it would be nice that scikit-optimize eventually expands beyond bayesian optimization to include more gradient-free optimization techniques.

(disclaimer: I'm not very familiar with bayesian optimization so please correct me if anything below is wrong).

Conceptually, NA like how it is presented in [Sambridge's paper](https://academic.oup.com/gji/article/138/2/479/596234#32272570) is actually very close to bayesian optimization as it also uses/refines a surrogate model for the objective function, which here is simply based on nearest neighbor interpolation (uniform probabilities are assumed inside each cell of the Voronoï tessellation). So I guess using ``sklearn.neighbors.KNearestNeiborsRegressor`` would just work here?

Where it departs from bayesian optimization in the strict sense is on how the next points are chosen: there is no optimization of such an acquisition function. Instead, the algorithm simply (more arbitrarily?) generates a fixed number ``ns`` of new points by uniform sampling within each of the ``nr`` “best” Voronoi cells.

Maybe there is a way to reformulate this in terms of an acquisition function so that it would be straightforward to implement it with skopt, I don’t know, would it make any sense? It looks like it would add unnecessary complexity. There might be some convergence issues when optimizing the acquisition function, too.

Another option would be to create a separate optimizer class that mirrors the API of `skopt.optimizer.Optimizer`. This seems perfectly feasible to me. My only (small) concern is to maintain such class so that it stays up-to-date with ``skopt``.

**Comment by betatim** [comment]
We discussed (or have interest in) adding other algorithms like for example in #636 and #625. There the issue is also a discussion around "do they fit into the current assumptions of skopt?" -> having a bit of a think what we can modify/loosen in terms of assumptions made in skopt to allow adding such other algorithms would be nice.

My impression was that there are some algorithms that only work with certain objective functions (like hyperband) and there are algorithms that need a special "acquisition function" (like Neighbourhood Algorithm). I say "acquisition function" because it would be a weird kind of acquisition function mostly to please the skopt conventions :) Off the top of my head a possible step towards supporting more exotic schemes of picking the next point(s) is to make our infrastructure so that acquisition functions are classes that encapsulate the optimisation of themselves and the optimizer just uses them (instead of currently where `Optimizer` decides how to optimise the acquisition function?

> Is `Optimizer` safe to use?

Yes. It is part of the public API so it is a bug if we break it without deprecation warning.

**Comment by benbovy** [comment]
> a possible step towards supporting more exotic schemes of picking the next point(s) is to make our infrastructure so that acquisition functions are classes that encapsulate the optimisation of themselves and the optimizer just uses them

Sounds like this would be very useful in this case! I'd be happy to start working on this as my first contribution to this project. Unfortunately I don't have much time for that right now.
